src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,183 @@
+            for (int i = size - 1; i > index; i--) {
+                nodeIndex = nodeIndex.previous;
+            }
+        }
+        return nodeIndex;
+    }
+
+    private void addToEmpty(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        first = newNode;
+        last = newNode;
+        size++;
+    }
+
+    public void addAsFirst(T value) {
+        if (size == 0) {
+            addToEmpty(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value, null, first);
+        first.previous = newNode;
+        first = newNode;
+        size++;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == 0) {
+            addToEmpty(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value, last, null);
+        last.next = newNode;
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("The index is out of range: " + index);
+        }
+        if (size == 0) {
+            addToEmpty(value);
+            return;
+        }
+        if (index == 0) {
+            addAsFirst(value);
+            return;
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value, getNode(index - 1), getNode(index + 1));
+        getNode(index - 1).next = newNode;
+        getNode(index + 1).previous = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T t : list) {
+            this.add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index is out of range: " + index);
+        }
+        return getNode(index).data;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index is out of range: " + index);
+        }
+        getNode(index).data = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index is out of range: " + index);
+        }
+
+        final Node<T> removedElement = getNode(index);
+        if (removedElement.previous == null) {
+            first = removedElement.next;
+        } else {
+            removedElement.previous.next = removedElement.next;
+            removedElement.previous = null;
+        }
+
+        if (removedElement.next == null) {
+            last = removedElement.previous;
+        } else {
+            removedElement.next.previous = removedElement.previous;
+            removedElement.next = null;
+        }
+
+        T dataToRemove = removedElement.data;
```

Comment:
we can write:
removedElement = null
and then remove removedElement.previous = null; and removedElement.next = null;